File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <stdlib.h>

int main(void)
{
    // Input credit card number from user
    string input = NULL;
    do
    {
        input = get_string("Number: ");
    }
    while (atol(input) < 0);

    int n = strlen(input);
    unsigned int copy[n + 2];
    unsigned int save[n + 1];
    copy[n + 2] = 0;

    // Luhn's Algorithm

    // Copy the input into 2 arrays of integers
    for (int i = 0; i <= n; i++)
    {
        copy[i] = input[i] - 48;
        save[i] = input[i] - 48;
    }
    for (int i = n - 2; i >= 0; i -= 2)
    {
        copy[i] = copy[i] * 2;
        if (copy[i] > 9)
        {
            copy[i] = ((copy[i] - (copy[i] % 10)) / 10) + (copy[i] % 10);
        }
    }
    for (int i = 0; i < n; i++)
    {
        copy[n + 2] += copy[i];
    }

    // Result of algorithm
    if ((copy[n + 2] % 10) == 0)
    {
        // American Express card number
        if (save[0] == 3 && (save[1] == 4 || save[1] == 7) && n == 15)
        {
            printf("AMEX\n");
            return 0;
        }
        // Visa card number
        else if (save[0] == 4 && (n == 13 || n == 16))
        {
            printf("VISA\n");
            return 0;
        }
        // Mastercard card number
        else if (save[0] == 5 && (save[1] == 1 || save[1] == 2 || save[1] == 3 || save[1] == 4 || save[1] == 5))
        {
            printf("MASTERCARD\n");
            return 0;
        }
        else
        {
            printf("INVALID\n");
            return 0;
        }
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
}

File: pset6/credit/credit.py
# Determine whether a provided credit card number is valid according to Luhn’s algorithm

from cs50 import get_int
from math import floor
from sys import exit


def main():

    # User input
    number = get_int("Number: ")
    
    # Call to function
    credit(number)


def credit(number):

    # Copy input
    copy = number

    # Calculate number length
    numlen = 0
    while copy > 0:
        copy = floor(copy / 10)
        numlen += 1
    
    # Check number length    
    if (numlen != 13 and numlen != 15 and numlen != 16):
        print("INVALID")
        exit(0)

    # Restore copy of input
    copy = number
    
    calc2 = 0
    copy *= 10
    
    # Luhn's algorithm
    i = 0
    while i < numlen:
        
        # Imagine that the last digit is the 0th, second-to-last is the 1st, and so forth.
        # (counting from right to left)
    
        # Calculation for EVEN numbered digits
        copy = floor((copy / 10))           # Proceed to next digit
        calc1 = copy 
        calc1 %= 10                         # Isolate last digit
        calc2 += calc1                      # Update total
        
        # Calculation for ODD numbered digits
        copy = floor((copy / 10))           # Proceed to next digit
        calc1 = copy 
        calc1 %= 10                         # Isolate last digit
        calc1 *= 2
        tmp1 = calc1 % 10 
        tmp2 = floor(calc1 / 10) % 10
        calc1 = tmp1 + tmp2
        calc2 += calc1                      # Update total
        i += 2
    
    # Restore copy of input
    copy = number
    
    # Isolate first 2 digits (left to right)
    while copy > 100:
        copy = floor(copy / 10)
    num2 = (copy % 10)                      # First digit
    num1 = floor((copy / 10)) % 10          # Second digit

    # Result of algorithm
    if (calc2 % 10) == 0:
        # American Express card number
        if num1 == 3 and (num2 == 4 or num2 == 7) and numlen == 15:
            print("AMEX")
            exit(0)

        # Visa card number
        elif num1 == 4 and (numlen == 13 or numlen == 16):
            print("VISA")
            exit(0)
        
        # Mastercard card number
        elif num1 == 5 and (num2 == 1 or num2 == 2 or num2 == 3 or num2 == 4 or num2 == 5):
            print("MASTERCARD")
            exit(0)
        else:
            print("INVALID")
            exit(0)
    else:    
        print("INVALID")
        exit(0)


main()
